docs: v4.0.0 cleanup (#1260)
exadel-inc/esl: pages/views/blogs/ESL v4.0.0.md

---
layout: content
name: ESL v4.0.0
title: ESL v4.0.0
tags: [blogs, draft]
date: 2022-10-01
containerCls: container markdown-container
---

## ESL Core

### Features:
  - `ESLEventListener`s  functionality introduced
  - `ESLMixinElement`s functionality introduced
  - `ESLBaseElement` extended API 
    - `ESLBaseElement.create` shortcut to create current custom element

### Structural changes:
  - Move attribute decorators to the `esl-utils`
  - All ESL modules migrated to `ESLEventListener`s
  - Event names now defined on a prototype level

### BREAKING CHANGES
  - `$$fire` no longer add 'esl:' prefix to the fired events

---

## ESL Media Query

### Features:
  - `ESLMediaQuery` now implements `EventTarget` interface
  - `ESLMediaRuleList` API reworked with `EventTarget` interface and new calculation strategy

### BREAKING CHANGES
  - `ESLMediaRuleList.parseTuple` arguments order changed

>   - `ESLMediaRule` no longer supports default marker, now "default" is equal "all" query

Reviewer: ```suggestion
  - `ESLMediaRule` no longer supports default marker, now "default" is equal to "all" query
```

  - `ESLMediaRuleList` observation callback signature changed, now it should be `EventListener`
  - `ESLMediaRuleList.prototype.default` removed as no longer default rules

>   - `ESLMediaRuleList` no longer fire events on rule change, now it's based on active valuer change

Reviewer: ```suggestion
  - `ESLMediaRuleList` no longer fires events on rule change, now it's based on active value change
```

  - `ESLMediaRuleList` now uses merging of all active rules to define result value, however you were still able to get the last active rule value

>   - `ESLMediaRuleList.prototype.active` now returns an array of active rules_

Reviewer 2: ```suggestion
  - `ESLMediaRuleList.prototype.active` now returns an array of active rules
```

  - simplify and extend `@prop` decorator API ([fd6ede3](https://github.com/exadel-inc/esl/commit/fd6ede34b0a3c7496083cb58aa9b726d4a692085))
  - add `skipOneRender` RAF utility ([ddc3227](https://github.com/exadel-inc/esl/commit/ddc322798e6f8cf447874896786fe6d368bbe5ef))
  - add ability to pass predicate to sequence finder ([dd8c3cb](https://github.com/exadel-inc/esl/commit/dd8c3cbb43ed529330c32459b9787949a0927a01))
  - add `@decorate` decorator to bind and decorate method of the class
  - `@attr` extended with ability to pass Serializer/Parser ([012eb83](https://github.com/exadel-inc/esl/commit/012eb83ebcbba8315a2766af090239774c21234e))
  - create dom html sanitize method ([004642f](https://github.com/exadel-inc/esl/commit/004642f866ac1b2ed5278084963288d8c2fd17e0))
  - create `extractValues` object utility ([8edd9e7](https://github.com/exadel-inc/esl/commit/8edd9e7b564b4ac78a4fc952ee58f76b6a54cf89))
  - create cumulative imports for esl-utils submodules ([2e9d6ad](https://github.com/exadel-inc/esl/commit/2e9d6ad8f327ede991feef4ebc7c76fdd55726af))

### BREAKING CHANGES

>   - `isMouseEvent`, `isTouchEvent` moved outside of `EventUtils`
>   - `normalizeTouchPoint` renamed to `touchPoint` and moved outside of `EventUtils`
>   - `normalizeCoordinates` removed `getOffsetPoint` introduced instead to be used in combination with `getTouchPoint`
>   - `@prop` signature changed `prop(value?: any, prototypeConfig: OverrideDecoratorConfig = {})`
>   - `ScrollUtils.lock` no longer accessible use `lockScroll(document.documentElement, {strategy: '...'})`
>   - `ScrollUtils.unlock` no longer accessible use `unlockScroll(document.documentElement, {strategy: '...'})`
>   - `ScrollUtils.lockRequest` no longer accessible use `lockScroll(document.documentElement, {strategy: '...', initiatior})`
>   - `ScrollUtils.unlockRequest` no longer accessible use `unlockScroll(document.documentElement, {strategy: '...', initiatior})`

Reviewer: ```suggestion
  - `isMouseEvent`, `isTouchEvent` moved outside of `EventUtils`
  - `normalizeTouchPoint` renamed to `touchPoint` and moved outside of `EventUtils`
  - `normalizeCoordinates` removed. Use `getOffsetPoint` in combination with `getTouchPoint` instead
  - `@prop` signature changed `prop(value?: any, prototypeConfig: OverrideDecoratorConfig = {})`
  - `ScrollUtils.lock` no longer accessible. Use `lockScroll(document.documentElement, {strategy: '...'})`
  - `ScrollUtils.unlock` no longer accessible. Use `unlockScroll(document.documentElement, {strategy: '...'})`
  - `ScrollUtils.lockRequest` no longer accessible. Use `lockScroll(document.documentElement, {strategy: '...', initiatior})`
  - `ScrollUtils.unlockRequest` no longer accessible. Use `unlockScroll(document.documentElement, {strategy: '...', initiatior})`
```

### Bugfixes
  - make attr decorators correctly strict typed
  - `deepMerge` primitive values merging improved, undefined arguments now ignored ([5b7b730](https://github.com/exadel-inc/esl/commit/5b7b73003acd4d29250af12603ced5d58c40aff6))
  - fix `@bind` decorator to save original function enumerable marker ([92c2086](https://github.com/exadel-inc/esl/commit/92c2086df610a5572143463302baf7aeea6174e9))
  - update `@decorate` decorator to allow to save context properly (on instance level) ([0ac1b0c](https://github.com/exadel-inc/esl/commit/0ac1b0cb9c90648ae1e0a18a16a936971c4b4ee9))

---

## ESL Toggleable and ESL Trigger

### Features:

>   - migrate `esl-togglable` and `esl-trigger` to ESLEventListeners

Reviewer 2: ```suggestion
  - migrate `esl-toggleable` and `esl-trigger` to ESLEventListeners
```

  - add support of dynamic `aria-label` for `esl-trigger` ([5c18841](https://github.com/exadel-inc/esl/commit/5c188418720ad4a972caa7450563a29268d8ec97))
  - add esc key event handler for `esl-trigger` ([3dab6da](https://github.com/exadel-inc/esl/commit/3dab6dad39259f7e920c9a108757271cb58a216d))
  - add `ignore-esc` attribute for `esl-trigger` ([605b715](https://github.com/exadel-inc/esl/commit/605b715a29ef1a852e24ce5120ca30d394fc19c1))

## ESL Panel and ESL Panel Group

### Features:
  - migrate `esl-panel` and `esl-panel-group` to ESLEventListeners
  - add `refresh-strategy` attribute ([#1156](https://github.com/exadel-inc/esl/issues/1156)) ([36027ad](https://github.com/exadel-inc/esl/commit/36027ad2574c0c33df1b6370484448a6889a647a))
  - add ability to control min/max open panels per media condition ([67ca2ba](https://github.com/exadel-inc/esl/commit/67ca2ba16e3fc4ee4300d1c79ca2e5d0da845af7))

### BREAKING CHANGES
  - `ESLPanelGroup.noCollapse` (with related attribute) renamed to `ESLPanelGroup.noAnimate` (`no-animate` attribute)
  - `ESLPanelGroup.shouldCollapse` renamed to `ESLPanelGroup.shouldAnimate`
  - `PanelActionParams.noCollapse` renamed to `PanelActionParams.noAnimate`

>   - `ESLPanelGroup` `view` attribute no longer supported

Reviewer: ```suggestion
  - `ESLPanelGroup` `view` attribute is no longer supported
```

  - `open` mode of `ESLPanelGroup` is no longer supported, it should be replaced with a `min-open-items="all"`

>   - `accordion-group="single"` attribute of `ESLPanelGroup` no longer supported, replaced with `max-open-items="1"`
>   - `accordion-group="single"` attribute of `ESLPanelGroup` no longer supported, replaced with `max-open-items="all"`

Reviewer: ```suggestion
  - `accordion-group="single"` attribute of `ESLPanelGroup` is no longer supported, replaced with `max-open-items="1"`
  - `accordion-group="single"` attribute of `ESLPanelGroup` is no longer supported, replaced with `max-open-items="all"`
```

  - components inherited from `ESLPanelGroup` should use `@listen({inherit: true})` for proper subscription

### Bugfixes

>   - add `capturedBy` and fix 'after:show' dispatch ([9bdc98c](https://github.com/exadel-inc/esl/commit/9bdc98c98a2926fe42c7cd15007c31a19171027c))

Reviewer 3: cosmetic:
```suggestion
  - add `capturedBy` and fix `after:show` dispatch ([9bdc98c](https://github.com/exadel-inc/esl/commit/9bdc98c98a2926fe42c7cd15007c31a19171027c))
```

  - fix `esl:before:hide` bubbling from uncontrolled toggleables ([9212b6b](https://github.com/exadel-inc/esl/commit/9212b6beacd97d39d0410a464598e9e2e0bdb1c2))
  - fix animation process capturing by `ESLPanel` component ([9a5b3a5](https://github.com/exadel-inc/esl/commit/9a5b3a576d5cb4dbd47d113d944932586155c7f6))
  - change `no-animate` API ([dde3500](https://github.com/exadel-inc/esl/commit/dde35009268a92bae92a45f00d11cb29edee98fa))

## ESL Tab and ESLTabs

### Features:

>   - migrate `esl-tab` and `esl-tabs` to ESLEventListeners

Reviewer: ```suggestion
  - Migrated `esl-tab` and `esl-tabs` to ESLEventListeners
```

### Bugfixes:

>   - fix alignment behavoir ([b5ebd66](https://github.com/exadel-inc/esl/commit/b5ebd668de7b4d1eea7311ae7d734c02abb537cf))

Reviewer 3: ```suggestion
  - fix alignment behavior ([b5ebd66](https://github.com/exadel-inc/esl/commit/b5ebd668de7b4d1eea7311ae7d734c02abb537cf))
```

### BREAKING CHANGES

>   - listeners extended from `ESLTab` should now use `@listen` annotation to work correctly
>   - listeners extended from `ESLTabs` should now use `@listen` annotation to work correctly

Reviewer: ```suggestion
  - Listeners extended from `ESLTab` should now use `@listen` annotation to work correctly
  - Listeners extended from `ESLTabs` should now use  `@listen` annotation to work correctly
```

## ESL Alert

### Features:
  - `esl-alert` migrated to ESLEventListeners, inner API updates

### BREAKING CHANGES

>   - listeners extended from `ESLAlert` should now use `@listen` annotation to work correctly

Reviewer: ```suggestion
  - Listeners extended from `ESLAlert` should now use `@listen` annotation to work correctly
```

>   - `eventNs` no longer supported, event now defined directly on the prototype level

Reviewer: ```suggestion
  - `eventNs` is no longer supported, event is now defined directly on the prototype level
```

## ESL Popup

### Features:
  - add configurable `rootMargin` for the popup activator observer ([5d647d6](https://github.com/exadel-inc/esl/commit/5d647d60673a65e7b27472c1b68a1e174fa40740))
  - add extended `offsetContainer` configuration ([63cbc0a](https://github.com/exadel-inc/esl/commit/63cbc0a996a8578549360dadaa3bb17c53884bfe))

### BREAKING CHANGES

>   - listeners extended from `ESLPopup` should now use `@listen` annotation to work correctly

Reviewer: ```suggestion
  - Listeners extended from `ESLPopup` should now use `@listen` annotation to work correctly
```

## ESL Footnotes

### Features:
  - migrate `esl-notes` and `esl-footnotes` to ESLEventListeners

>   - add configurable `intersectionMargin  for the note tooltip activator observer ([b9b1599](https://github.com/exadel-inc/esl/commit/b9b159942e3534965f43cac94c140d5680308548))

Reviewer: ```suggestion
  - add configurable `intersectionMargin`  for the note tooltip activator observer ([b9b1599](https://github.com/exadel-inc/esl/commit/b9b159942e3534965f43cac94c140d5680308548))
```

### BREAKING CHANGES
  - listeners extended from `ESLNote` should now use `@listen` annotation to work correctly
  - listeners extended from `ESLFootnote` should now use `@listen` annotation to work correctly

>   - `eventNs` no longer supported, event now defined directly on the prototype level

Reviewer: ```suggestion
  - `eventNs` is no longer supported, event is now defined directly on the prototype level
```

>   - listeners extended from `ESLNote` should now use `@listen` annotation to work correctly
>   - listeners extended from `ESLFootnote` should now use `@listen` annotation to work correctly
>   - `eventNs` no longer supported, event now defined directly on the prototype level

Reviewer: ```suggestion
  - Listeners extended from `ESLNote` should now use `@listen` annotation to work correctly
  - Listeners extended from `ESLFootnote` should now use `@listen` annotation to work correctly
  - `eventNs` is no longer supported, event is now defined directly on the prototype level
```

### Structural changes:
  - Move attribute decorators to the `esl-utils`
  - All ESL modules migrated to `ESLEventListener`s
  - Event names now defined on a prototype level

### BREAKING CHANGES

>   - `$$fire` no longer add 'esl:' prefix to the fired events

Reviewer 4: ```suggestion
  - `$$fire` no longer adds 'esl:' prefix to the fired events
```

  - fix `esl:before:hide` bubbling from uncontrolled toggleables ([9212b6b](https://github.com/exadel-inc/esl/commit/9212b6beacd97d39d0410a464598e9e2e0bdb1c2))
  - fix animation process capturing by `ESLPanel` component ([9a5b3a5](https://github.com/exadel-inc/esl/commit/9a5b3a576d5cb4dbd47d113d944932586155c7f6))
  - change `no-animate` API ([dde3500](https://github.com/exadel-inc/esl/commit/dde35009268a92bae92a45f00d11cb29edee98fa))

## ESL Tab and ESLTabs

### Features:
  - migrate `esl-tab` and `esl-tabs` to ESLEventListeners

### Bugfixes:
  - fix alignment behavoir ([b5ebd66](https://github.com/exadel-inc/esl/commit/b5ebd668de7b4d1eea7311ae7d734c02abb537cf))

### BREAKING CHANGES
  - listeners extended from `ESLTab` should now use `@listen` annotation to work correctly
  - listeners extended from `ESLTabs` should now use `@listen` annotation to work correctly

## ESL Alert

### Features:
  - `esl-alert` migrated to ESLEventListeners, inner API updates

### BREAKING CHANGES
  - listeners extended from `ESLAlert` should now use `@listen` annotation to work correctly

>   - `eventNs` no longer supported, event now defined directly on the prototype level

Reviewer 4: ```suggestion
  - `eventNs` is no longer supported, event now defines directly on the prototype level
```

  - add configurable `rootMargin` for the popup activator observer ([5d647d6](https://github.com/exadel-inc/esl/commit/5d647d60673a65e7b27472c1b68a1e174fa40740))
  - add extended `offsetContainer` configuration ([63cbc0a](https://github.com/exadel-inc/esl/commit/63cbc0a996a8578549360dadaa3bb17c53884bfe))

### BREAKING CHANGES
  - listeners extended from `ESLPopup` should now use `@listen` annotation to work correctly

## ESL Footnotes

### Features:
  - migrate `esl-notes` and `esl-footnotes` to ESLEventListeners
  - add configurable `intersectionMargin  for the note tooltip activator observer ([b9b1599](https://github.com/exadel-inc/esl/commit/b9b159942e3534965f43cac94c140d5680308548))

### BREAKING CHANGES
  - listeners extended from `ESLNote` should now use `@listen` annotation to work correctly
  - listeners extended from `ESLFootnote` should now use `@listen` annotation to work correctly

>   - `eventNs` no longer supported, event now defined directly on the prototype level

Reviewer 4: ```suggestion
  - `eventNs` is no longer supported, event now defines directly on the prototype level
```
